### PR TITLE
[Seq] Use a new pass generation mechanism

### PR DESCRIPTION
**Pull request by uenoku**
This applies a new pass generation method(ref:https://github.com/llvm/circt/issues/3962) to seq dialect passes, that automatically defines a struct of options (e.g. `LowerSeqFIRRTLToSVOptions`) and creates a pass constructor to take the option.

### CODE REVIEW

File: lib/Dialect/Seq/Transforms/PassDetails.h [hunk 1]
```diff
@@ -17,12 +17,15 @@
...
 #include "circt/Dialect/SV/SVDialect.h"
 #include "circt/Dialect/Seq/SeqOps.h"
+#include "circt/Dialect/Seq/SeqPasses.h"
 #include "mlir/Pass/Pass.h"
 
 namespace circt {
 namespace seq {
 
-#define GEN_PASS_CLASSES
+#define GEN_PASS_DEF_LOWERSEQFIRRTLTOSV
```
Comment (darthscsi) [inline lib/Dialect/Seq/Transforms/PassDetails.h:26]: Is there a generic "lower all pass_def" so we don't have to define each pass' define?

File: lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp [hunk 2]
```diff
@@ -31,12 +31,14 @@ using namespace circt;
 using namespace seq;
 
 namespace {
-struct SeqToSVPass : public LowerSeqToSVBase<SeqToSVPass> {
+struct SeqToSVPass : public circt::seq::impl::LowerSeqToSVBase<SeqToSVPass> {
```
Comment (darthscsi) [inline lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp:34]: circt::seq:: seems redundant given the using declaration.

File: lib/Dialect/Seq/Transforms/LowerSeqHLMem.cpp [hunk 3]
```diff
@@ -148,7 +148,8 @@ struct SimpleBehavioralMemoryLowering
   }
 };
 
-struct LowerSeqHLMemPass : public LowerSeqHLMemBase<LowerSeqHLMemPass> {
+struct LowerSeqHLMemPass
+    : public circt::seq::impl::LowerSeqHLMemBase<LowerSeqHLMemPass> {
```
Comment (darthscsi) [inline lib/Dialect/Seq/Transforms/LowerSeqHLMem.cpp:152]: circt::seq:: seems redundant given the using declaration.

File: lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp [hunk 4]
```diff
@@ -31,12 +31,14 @@ using namespace circt;
...
   void runOnOperation() override;
 };
-struct SeqFIRRTLToSVPass : public LowerSeqFIRRTLToSVBase<SeqFIRRTLToSVPass> {
+struct SeqFIRRTLToSVPass
+    : public circt::seq::impl::LowerSeqFIRRTLToSVBase<SeqFIRRTLToSVPass> {
```
Comment (darthscsi) [inline lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp:38]: ditto

File: lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp [hunk 5]
```diff
@@ -520,10 +522,7 @@ std::unique_ptr<Pass> circt::seq::createSeqLowerToSVPass() {
...
 }
 
-std::unique_ptr<Pass>
-circt::seq::createSeqFIRRTLLowerToSVPass(bool disableRegRandomization) {
-  auto pass = std::make_unique<SeqFIRRTLToSVPass>();
-  if (disableRegRandomization)
-    pass->disableRegRandomization = disableRegRandomization;
-  return pass;
+std::unique_ptr<Pass> circt::seq::createSeqFIRRTLLowerToSVPass(
+    const circt::seq::LowerSeqFIRRTLToSVOptions &options) {
```
Comment (darthscsi) [inline lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp:526]: The namespaces on the options might be unnecessary.

**Review by darthscsi — APPROVED** [review]

File: lib/Dialect/Seq/Transforms/PassDetails.h (hunk 1 above)
Comment (uenoku) [reply lib/Dialect/Seq/Transforms/PassDetails.h:26]: Currently there is no such macro but it would be definitely useful so i'll open MLIR issue.

**Pull request merged**